addons/web/static/src/webclient/company_service.js

```diff
@@ -114,14 +114,46 @@ export const companyService = {
                     nextAvailableCompanies.add(companyId);
                 }
             },
-            logNextCompanies() {
+            async logNextCompanies() {
                 const next = nextAvailableCompanies.length
                     ? nextAvailableCompanies
                     : [allowedCompanyIds[0]];
-                router.pushState(
-                    { cids: formatCompanyIds(next, CIDS_HASH_SEPARATOR) },
-                    { lock: true }
-                );
+
+                const currentModel = router.current.hash.model;
+                const currentId = router.current.hash.id;
+                const currentMenu = router.current.hash.menu_id;
+
+                let stayOnCurrentPage = true;
+
+                if (currentModel && currentId)
+                {
+                    stayOnCurrentPage = await env.services.rpc(
+                        "/web/webclient/check_record_company",
+                        {
+                            model: currentModel,
+                            res_id: currentId,
+                            company_ids: next,
+                        },
+                    );
+                }
+
+                if (stayOnCurrentPage)
+                {
+                    router.pushState(
```

Comment:
the session's context wouldn't be altered for later navigation, would it ?

Reply:
I'm not sure, need to double check !

Reply:
then it is a problem. The user would not have effectively changed companies if the context did not change.

Flow to try (and if this is to be continued it needs a proper test, I guess as a tour)
- be on invoice
- change companies that would result in your method return false
- go back to the invoices list.

What a users expects to see is the list of invoices of the new company.
I don't see it in your code.

Also, to check functionally, in odoo until now, changing companies meant reloading: there was a feedback
With your PR, the only feedback is the url, which seems insufficient